lib/cartopy/mpl/gridliner.py

```diff
@@ -327,58 +338,257 @@ def _draw_gridliner(self, nx=None, ny=None, background_patch=None):
+        angle = np.arctan2(y0-y1, x0-x1) * 180 / np.pi
+        kw, loc = self._segment_angle_to_text_specs(angle)
+        return kw, angle, loc
+
+    def _segment_angle_to_text_specs(self, angle):
+        """Get appropriate kwargs for a given direction angle"""
+        kw, loc = _text_angle_to_specs_(angle)
+        if not self.rotate_labels:
+            angle = {'top': 90., 'right': 0.,
+                     'bottom': -90., 'left': 180.}[loc]
+            del kw['rotation']
+
+        xpadding = (self.xpadding if self.xpadding is not None
+                    else matplotlib.rc_params['xtick.major.pad'])
+        ypadding = (self.ypadding if self.ypadding is not None
+                    else matplotlib.rc_params['ytick.major.pad'])
+        dx = ypadding * np.cos(angle * np.pi / 180)
+        dy = xpadding * np.sin(angle * np.pi / 180)
+        transform = mtrans.offset_copy(self.axes.transData, self.axes.figure,
+                                       x=dx, y=dy, units='dots')
+        kw.update(transform=transform)
+
+        return kw, loc
+
+    def _update_labels_visibility(self, renderer):
+        """Update the visibility of each plotted label
+
+        The following rules apply:
+
+        - Labels are plotted and checked by order of priority,
+          with a high priority for longitude labels at the bottom and
+          top of the map, and the reverse for latitude labels.
+        - A label must not overlap another label marked as visible.
+        - A label must not overlap the map boundary.
+        - When a label is about to be hidden, other angles are tried in the
+          absolute given limit of max_delta_angle by increments of delta_angle
+          of difference from the original angle.
 
-        if self.xlabels_bottom or self.xlabels_top:
-            self._assert_can_draw_ticks()
-            self.xformatter.set_locs(x_label_points)
-            for x in x_label_points:
-                if self.xlabels_bottom:
-                    self._add_gridline_label(x, axis='x', upper_end=False)
-                if self.xlabels_top:
-                    self._add_gridline_label(x, axis='x', upper_end=True)
-
-        if self.ylabels_left or self.ylabels_right:
-            self._assert_can_draw_ticks()
-            self.yformatter.set_locs(y_label_points)
-            for y in y_label_points:
-                if self.ylabels_left:
-                    self._add_gridline_label(y, axis='y', upper_end=False)
-                if self.ylabels_right:
-                    self._add_gridline_label(y, axis='y', upper_end=True)
+        """
+        if renderer is None or not self._labels:
+            return
+        paths = []
+        outline_path = None
+        delta_angle = 22.5
+        max_delta_angle = 45
+        axes_children = self.axes.get_children()
+        for priority, artist in self._labels:
+
+            if artist not in axes_children:
+                warn('The labels if this gridliner do not belongs'
```

Comment:
A couple of little typos here; `'The labels if this gridliner do not belongs'` -> `'The labels of this gridliner do not belong'`

Reply:
Fixed